Add UnionType#resolve_type for more ganular control in unions

colepatrickturner opened the pull request:
## Purpose
This feature allows a union type to resolve to more specific types when handling a value. This is useful when working with objects that may fit into multiple types. When executing the query
will call `resolve_type(value, field_ctx)` on either Unions or Interfaces.

Union type's call can be definable via `resolve_type ->(value, ctx) { }`. Interface type's call returns `ctx.query.resolve_type(value)` which is the default implementation for resolving both types.

### Tests:
- Added a new test that raises an error if a returned type is not within possible_types
- Added an implementation in the dummy schema, though I would consider altering this by adding a new type derived from MilkType (for example KefirType) in which it is re-casted as a MilkType - but this would alter a significant portion of other tests.

### Guides
- I have not updated any guides or documentation on the static site. If there are any I should do for this PR please advise.

```
831 tests, 4720 assertions, 0 failures, 0 errors, 0 skips
Running RuboCop...
Inspecting 340 files
....................................................................................................................................................................................................................................................................................................................................................

340 files inspected, no offenses detected
```

Edit:

Resolves #529

rmosolgo commented on lib/graphql/union_type.rb:
```diff
@@ -24,19 +24,21 @@ module GraphQL
       super
       @clean_possible_types = nil
       @dirty_possible_types = other.dirty_possible_types.dup
+      @resolve_type_proc = other.resolve_type_proc
```
I think this is the default behavior of `#initialize_copy`, to assign each ivar of the original to to the new copy, so we can remove this line and let `super` handle it.

colepatrickturner replied:
Ok! Will remove

rmosolgo commented on lib/graphql/union_type.rb:
```diff
@@ -64,8 +66,25 @@ def possible_types
+          raise(NotImplementedError, "Unrecognized possible type kind: #{resolved_type}")
+        end
+
+        return resolved_type
```
It's possible to [hide types and fields](http://graphql-ruby.org/schema/limiting_visibility.html) for individual queries. Should we do anything special here if the user-provided `@resolve_type_proc` returns a type which _ought_ to be hidden for this query?

rmosolgo commented on lib/graphql/execution/execute.rb:
```diff
@@ -210,7 +210,7 @@ def resolve_value(owner, parent_type, field_defn, field_type, value, selection,
             when GraphQL::TypeKinds::UNION, GraphQL::TypeKinds::INTERFACE
               query = field_ctx.query
-              resolved_type = query.resolve_type(value)
+              resolved_type = field_type.resolve_type(value, field_ctx)
```
:+1: Seems right! Is this change covered by a corresponding change in the tests, to ensure that we _keep_ using the type's `resolve_type` in future versions?

colepatrickturner replied:
I don't believe so, I'll add a new one for that.

colepatrickturner commented:
> We should add a specific test for the new feature (that objects of the same Ruby class can resolve to different types at different times).

I have rebased in a test and some changes to the schema that don't change much while trying to use the same schema.

- Restructured `COW` from OpenStruct to Struct for the purpose of being usable in a Union
I was worried about potential conflict here changing from OpenStruct to Struct, because previous tests used `COW` in a `GraphQL::Function` to specify a type. However that does not affect the use of functions to derive type within testing.

- Added `Goat`, `GOATS` and `COWS`, `AnimalUnion` and `AnimalAsCowUnion`
These are used in the dummy schema derive types within unions.
`AnimalUnion` tests the default implementation using the same types that `AnimalAsCowUnion` will type-cast into `CowType`.

- Added `Query->allAnimal` and `Query->allAnimalAsCow`
`allAnimal` returns both the Cow and Goat while `allAnimalAsCow` will typecast a goat into a cow.

I hope I covered the tests per your request. My intention was to use the same schema because I believe this change to be imperative to testing a whole schema to ensure it's compatibility between schema-level and union-level resolution. Thanks for your review and time :)

rmosolgo commented:
This looks great! Nice work coercing that wacky schema to suit your needs :)

colepatrickturner commented:
Thank you 🙏

broskoski commented:
👏 Nice work, thank you!

akaKuruma commented:
Thank you! 🍻
